ci/L0_backend_vllm/metrics_test/vllm_metrics_test.py

```diff
@@ -0,0 +1,136 @@
+
+import os
+import re
+import sys
+import unittest
+from functools import partial
+
+import requests
+import tritonclient.grpc as grpcclient
+from tritonclient.utils import *
+
+sys.path.append("../../common")
+from test_util import TestResultCollector, UserData, callback, create_vllm_request
+
+
+class VLLMTritonMetricsTest(TestResultCollector):
+    def setUp(self):
+        self.triton_client = grpcclient.InferenceServerClient(url="localhost:8001")
+        self.tritonserver_ipaddr = os.environ.get("TRITONSERVER_IPADDR", "localhost")
+        self.vllm_model_name = "vllm_opt"
+        self.prompts = [
+            "The most dangerous animal is",
+            "The capital of France is",
+            "The future of AI is",
+        ]
+        self.sampling_parameters = {"temperature": "0", "top_p": "1"}
+
+    def get_metrics(self):
+        """
+        Store vllm metrics in a dictionary.
+        """
+        r = requests.get(f"http://{self.tritonserver_ipaddr}:8002/metrics")
+        r.raise_for_status()
+
+        # Regular expression to match the pattern
+        pattern = r"^(vllm:[^ {]+)(?:{.*})? ([0-9.-]+)$"
+        vllm_dict = {}
+
+        # Find all matches in the text
+        matches = re.findall(pattern, r.text, re.MULTILINE)
+
+        for match in matches:
+            key, value = match
+            vllm_dict[key] = float(value) if "." in value else int(value)
+
+        return vllm_dict
+
+    def vllm_infer(
+        self,
+        prompts,
+        sampling_parameters,
+        model_name,
+    ):
+        """
+        Helper function to send async stream infer requests to vLLM.
+        """
+        user_data = UserData()
+        number_of_vllm_reqs = len(prompts)
+
+        self.triton_client.start_stream(callback=partial(callback, user_data))
+        for i in range(number_of_vllm_reqs):
+            request_data = create_vllm_request(
+                prompts[i],
+                i,
+                False,
+                sampling_parameters,
+                model_name,
+                True,
+            )
+            self.triton_client.async_stream_infer(
+                model_name=model_name,
+                inputs=request_data["inputs"],
+                request_id=request_data["request_id"],
+                outputs=request_data["outputs"],
+                parameters=sampling_parameters,
+            )
+
+        for _ in range(number_of_vllm_reqs):
+            result = user_data._completed_requests.get()
+            if type(result) is InferenceServerException:
+                print(result.message())
+            self.assertIsNot(type(result), InferenceServerException, str(result))
+
+            output = result.as_numpy("text_output")
+            self.assertIsNotNone(output, "`text_output` should not be None")
+
+        self.triton_client.stop_stream()
+
+    def test_vllm_metrics(self):
+        # All vLLM metrics from tritonserver
```

Comment:
Clarify `All`. For anyone else, who would need to work on this test later, they should be fully aware of what this test tests. Something like: this test makes sure that counter metrics are collected properly.

BTW, how do we test `gauge` metrics?

Reply:
The test was updated to allow new metrics to be added.
We don't care about gauge metrics for now since there is none.
Most guage metrics have non-zero values in the middle of an inference but are set to zero when it finishes. We need to come up with a solution that captures all values have been set to gauge metrics.

Reply:
> We don't care about gauge metrics for now since there is none.
There is none, meaning?